devtools/etdump/buffer_data_sink.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <executorch/devtools/etdump/buffer_data_sink.h>
+#include <executorch/devtools/etdump/utils.h>
+
+using ::executorch::runtime::Error;
+using ::executorch::runtime::Result;
+
+namespace executorch {
+namespace etdump {
+
+Result<size_t> BufferDataSink::write(const void* ptr, size_t length) {
+  if (length == 0) {
+    return static_cast<size_t>(-1);
+  }
+  uint8_t* offset_ptr =
+      internal::align_pointer(debug_buffer_.data() + offset_, 64);
+
+  // Zero out the padding between data blobs.
+  size_t n_zero_pad = offset_ptr - debug_buffer_.data() - offset_;
+  memset(debug_buffer_.data() + offset_, 0, n_zero_pad);
+
+  offset_ = (offset_ptr - debug_buffer_.data()) + length;
+
+  // Raise access error if offset_ is out of range.
+  if (offset_ > debug_buffer_.size()) {
+    return Error::OutOfResources;
+  }
```

Comment:
This check should happen before mutating the class in any way (zeroing padding, updating offset_). I.e., a failure case should leave the state untouched.

E.g., someone might say `write(ptr, 1MB)` which fails, then `write(ptr 1kB)` which succeeds. The offset shouldn't move, and the buffer shouldn't be modified, when the first call fails.